src/openclaw_pipeline/packs/research_tech/truth_projection.py

```
@@ -1,16 +1,242 @@
+
+def _subject_key(claim_text: str) -> str:
+    lowered = claim_text.strip().lower()
+    lowered = re.sub(r"\s+", " ", lowered)
+    for marker in (" supports ", " does not support ", " is ", " are ", " has ", " have "):
+        if marker in lowered:
+            return lowered.split(marker, 1)[0].strip()
+    return lowered.split(".", 1)[0].strip()
+
+
+def _contradiction_id_for_subject(subject: str) -> str:
+    return f"contradiction::{hashlib.sha1(subject.encode('utf-8')).hexdigest()[:12]}"
```

Ruff (0.15.10)

[error] Probable use of insecure hash functions in `hashlib`: `sha1`

(S324)